nl_server/query_util.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Query related helpers"""
+
+from dataclasses import dataclass
+import logging
+import re
+from typing import List
+
+from server.lib.nl import constants
+from server.lib.nl import utils
+
+# TODO: decouple words removal from detected attributes. Today, the removal
+# blanket removes anything that matches, including the various attribute/
+# classification triggers and contained_in place types (and their plurals).
+# This may not always be the best thing to do.
+ALL_STOP_WORDS = utils.combine_stop_words()
+
+_MAX_SVS = 4
+
+# Use comma, "vs.", semi-colon, "and", ampersand as delimiters.
+_REGEX_DELIMITERS = r',|vs|;|and|&'
+# Regex to extract out substrings within double quotes.
+_REGEX_QUOTED_STRING = r'"([^"]+)"'
+
+
+def _remove_punctuations(s):
+  s = s.replace('\'s', '')
+  s = re.sub(r'[^\w\s]', ' ', s)
+  return " ".join(s.split())
+
+
+# A specific way a query has been split.
+@dataclass
+class QuerySplit:
+  # Parts of a query
+  parts: List[str]
+
+
+# Different combinations of splits for a given nsplits.
+@dataclass
+class QuerySet:
+  # The number of splits
+  nsplits: int
+  # Was it based on delimiter?
+  delim_based: bool
+  # Different combinations of a query
+  combinations: List[QuerySplit]
+
+
+#
```

Comment:
It would be great to document this with some examples. With the recursive calls inside this function, highly recommended extensive documentation because it is fairly hairy to figure out what's going on. I also recommend more comments to help step line by line.